pants-plugins/internal_plugins/test_lockfile_fixtures/rules.py

```diff
@@ -0,0 +1,270 @@
+    if not mark:
+        continue
+
+    path = Path(item.path).relative_to(cwd)
+
+    output.append({
+        "kwargs": mark.kwargs,
+        "test_file_path": str(path),
+    })
+
+with open("tests.json", "w") as f:
+    f.write(json.dumps(output))
+"""
+
+
+@dataclass(frozen=True)
+class JVMLockfileFixtureConfig:
+    definition: JVMLockfileFixtureDefinition
+    test_file_path: str
+
+
+class CollectedJVMLockfileFixtureConfigs(DeduplicatedCollection[JVMLockfileFixtureConfig]):
+    pass
+
+
+@dataclass(frozen=True)
+class RenderedJVMLockfileFixture:
+    content: bytes
+    path: str
+
+
+class RenderedJVMLockfileFixtures(DeduplicatedCollection[RenderedJVMLockfileFixture]):
+    pass
+
+
+@dataclass(frozen=True)
+class CollectFixtureConfigsRequest:
+    pass
+
+
+@rule
+async def collect_fixture_configs(
+    _request: CollectFixtureConfigsRequest,
+    pytest: PyTest,
+    python_setup: PythonSetup,
+    test_extra_env: TestExtraEnv,
+) -> CollectedJVMLockfileFixtureConfigs:
+    specs_parser = SpecsParser()
+    specs = specs_parser.parse_specs(["src/python/pants::"])
+    targets = await Get(Targets, AddressSpecs, specs.address_specs)
+    addresses = [tgt.address for tgt in targets]
+    transitive_targets = await Get(TransitiveTargets, TransitiveTargetsRequest(addresses))
+    all_targets = transitive_targets.closure
+
+    interpreter_constraints = InterpreterConstraints.create_from_targets(all_targets, python_setup)
+
+    requirements_pex_get = Get(Pex, RequirementsPexRequest(addresses))
+    pytest_pex_get = Get(
+        Pex,
+        PexRequest(
+            output_filename="pytest.pex",
+            requirements=pytest.pex_requirements(),
+            interpreter_constraints=interpreter_constraints,
+            internal_only=True,
+        ),
+    )
+
+    prepared_sources_get = Get(
+        PythonSourceFiles, PythonSourceFilesRequest(all_targets, include_files=True)
+    )
+
+    (pytest_pex, requirements_pex, prepared_sources,) = await MultiGet(
+        pytest_pex_get,
+        requirements_pex_get,
+        prepared_sources_get,
+    )
+
+    script_content = FileContent(
+        path="collect-fixtures.py", content=COLLECTION_SCRIPT.encode(), is_executable=True
+    )
+    script_digest = await Get(Digest, CreateDigest([script_content]))
+
+    pytest_runner_pex_get = Get(
+        VenvPex,
+        PexRequest(
+            output_filename="pytest_runner.pex",
+            interpreter_constraints=interpreter_constraints,
+            main=EntryPoint(PurePath(script_content.path).stem),
+            sources=script_digest,
+            internal_only=True,
+            pex_path=[
+                pytest_pex,
+                requirements_pex,
+            ],
+        ),
+    )
+    config_file_dirs = list(group_by_dir(prepared_sources.source_files.files).keys())
+    config_files_get = Get(
+        ConfigFiles,
+        ConfigFilesRequest,
+        pytest.config_request(config_file_dirs),
+    )
```

Comment:
Needing to copy things like config collection makes me think that maybe the body of the pytest runner `@rule` should maybe be extracted into a `@rule_helper` for reuse here?